prov/rxm/src/rxm_cq.c

```diff
@@ -567,32 +567,63 @@ static ssize_t rxm_cq_write_error(struct fid_cq *msg_cq,
 	return ofi_cq_write_error(util_cq, &err_entry);
 }
 
-void rxm_cq_progress(struct rxm_ep *rxm_ep)
+#define RXM_CQ_READ(rxm_ep, err_again, err_other)		\
+	do {							\
+		ret = fi_cq_read((rxm_ep)->msg_cq, &comp, 1);	\
+		if (ret == -FI_EAGAIN)				\
+			goto err_again;				\
+		if (OFI_UNLIKELY(ret < 0))			\
+			goto err_other;				\
+		if (ret) {					\
+			comp_read += (size_t)ret;		\
+			ret = rxm_cq_handle_comp(rxm_ep, &comp);\
+			if (OFI_UNLIKELY(ret))			\
+				goto err_other;			\
+		}						\
+	} while (0)
```

Comment:
I'm definitely not a fan of this coding structure style.  It ends up being extremely difficult to read and maintain.  There's got to be a better option here, especially since the progress and progress_ex calls are nearly identical.

Reply:
I agree with you. Ok, I'll look for a good way here.